src/phase0/state_transition/epoch_processing.rs

```diff
@@ -444,6 +526,437 @@ pub fn process_participation_record_updates<
+    >,
+    context: &Context,
+) -> Result<(Vec<Gwei>, Vec<Gwei>), Error> {
+    // Return inactivity reward/penalty deltas for each validator.
+    let previous_epoch = get_previous_epoch(state, context);
+    let validator_count = state.validators.len();
+    let mut penalties = vec![0; validator_count];
+    if is_in_inactivity_leak(state, context) {
+        let matching_target_attestations =
+            get_matching_target_attestations(state, previous_epoch, context)?
+                .cloned()
+                .collect::<Vec<_>>();
+        let matching_target_attesting_indices = get_unslashed_attesting_indices(
+            state,
+            matching_target_attestations.as_slice(),
+            context,
+        )?;
+        for i in get_eligible_validator_indices(state, context) {
+            // If validator is performing optimally this cancels all rewards for a neutral balance
+            let base_reward = get_base_reward(state, i, context)?;
+            penalties[i] +=
+                BASE_REWARDS_PER_EPOCH * base_reward - get_proposer_reward(state, i, context)?;
+            if !matching_target_attesting_indices.contains(&i) {
+                let effective_balance = state.validators[i].effective_balance;
+                penalties[i] += effective_balance * get_finality_delay(state, context)
+                    / context.inactivity_penalty_quotient;
+            }
+        }
+    }
+    // No rewards associated with inactivity penalties
+    let rewards = vec![0; validator_count];
+    Ok((rewards, penalties))
+}
+
+pub fn get_attestation_deltas<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: &BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    context: &Context,
+) -> Result<(Vec<Gwei>, Vec<Gwei>), Error> {
+    // Return attestation reward/penalty deltas for each validator.
+    let (source_rewards, source_penalties) = get_source_deltas(state, context)?;
+    let (target_rewards, target_penalties) = get_target_deltas(state, context)?;
+    let (head_rewards, head_penalties) = get_head_deltas(state, context)?;
+    let (inclusion_delay_rewards, _) = get_inclusion_delay_deltas(state, context)?;
+    let (_, inactivity_penalties) = get_inactivity_penalty_deltas(state, context)?;
+
+    let validator_count = state.validators.len();
+    let mut rewards = vec![0; validator_count];
+    for i in 0..validator_count {
```

Comment:
Wasn't sure if this was the best approach or doing:
- `rewards.map(...)` instead of the `for` loop
- Making `rewards` an empty `Vec` and then pushing into it during the `for` loop

Same obv for below

Reply:
whatever gets correct code written in the shortest possible time :)

one thing to consider is that we will likely come optimize this and a big source of speed ups will be eliminating the "iter over all validators" pattern that ends up everywhere in the "naive" spec